File: scripts/fix_admonitions.py
#!/usr/bin/env python3
"""
Fix unsupported admonition types for Typst export.
Converts tip, important, and generic admonition blocks to supported types.
"""
import re
import sys
from pathlib import Path

def fix_admonitions_in_file(filepath):
    """Fix admonitions in a single markdown file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original_content = content
    changes_made = []

    # Pattern to match admonition blocks
    # Matches: ```{admonition} Title\n or ```{tip}\n or ```{important}\n

    # Fix generic admonitions with custom titles
    # Convert ```{admonition} Custom Title to ```{note} Custom Title
    pattern_admonition = r'```\{admonition\}([^\n]*)\n'
    if re.search(pattern_admonition, content):
        content = re.sub(pattern_admonition, r'```{note}\1\n', content)
        changes_made.append("admonition → note")

    # Fix tip admonitions - convert to note with "Tip:" prefix
    pattern_tip = r'```\{tip\}\n'
    if re.search(pattern_tip, content):
        # For tip blocks, we need to add "Tip" as the title
        # This is a bit tricky as we need to handle the content properly
        # For now, convert to note and rely on context
        content = re.sub(pattern_tip, '```{note}\n:class: tip\n\n**Tip:** ', content)
        changes_made.append("tip → note with Tip prefix")

    # Fix important admonitions - convert to warning with "Important:" prefix
    pattern_important = r'```\{important\}\n'
    if re.search(pattern_important, content):
        content = re.sub(pattern_important, '```{warning}\n:class: important\n\n**Important:** ', content)
        changes_made.append("important → warning with Important prefix")

    if content != original_content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return True, changes_made
    return False, []

def main():
    """Process all markdown files in the content directory."""
    content_dir = Path('/home/user/opticsTextbook/content')

    if not content_dir.exists():
        print(f"Error: Content directory not found: {content_dir}")
        sys.exit(1)

    markdown_files = list(content_dir.rglob('*.md'))

    print(f"Found {len(markdown_files)} markdown files")
    print("=" * 60)

    total_changed = 0
    for filepath in markdown_files:
        changed, changes = fix_admonitions_in_file(filepath)
        if changed:
            total_changed += 1
            rel_path = filepath.relative_to(content_dir)
            print(f"✓ {rel_path}")
            for change in changes:
                print(f"  - {change}")

    print("=" * 60)
    print(f"Modified {total_changed} files")

    if total_changed == 0:
        print("No changes needed!")
    else:
        print("\nAdmonitions have been converted to supported types.")
        print("Run 'npx myst build --typst' to verify the fixes.")

if __name__ == '__main__':
    main()

File: scripts/fix_admonitions.js
#!/usr/bin/env node
/**
 * Fix unsupported admonition types for Typst export.
 * Converts tip, important, and generic admonition blocks to supported types.
 *
 * Status: HISTORICAL - One-time migration script
 *     Used during initial Typst export setup (December 2025).
 *     Retained for reference and potential future use if new admonition
 *     types are introduced that need conversion.
 */

const fs = require('fs');
const path = require('path');

/**
 * Fix admonitions in a single markdown file.
 *
 * @param {string} filepath - Path to markdown file
 * @returns {Array} [changed, changesMade]
 */
function fixAdmonitionsInFile(filepath) {
  const content = fs.readFileSync(filepath, 'utf8');
  const originalContent = content;
  let modifiedContent = content;
  const changesMade = [];

  // Fix generic admonitions with custom titles
  // Convert ```{admonition} Custom Title to ```{note} Custom Title
  const patternAdmonition = /```\{admonition\}([^\n]*)\n/g;
  if (patternAdmonition.test(content)) {
    modifiedContent = modifiedContent.replace(patternAdmonition, '```{note}$1\n');
    changesMade.push('admonition → note');
  }

  // Fix tip admonitions - convert to note with "Tip:" prefix
  const patternTip = /```\{tip\}\n/g;
  if (patternTip.test(content)) {
    modifiedContent = modifiedContent.replace(patternTip, '```{note}\n:class: tip\n\n**Tip:** ');
    changesMade.push('tip → note with Tip prefix');
  }

  // Fix important admonitions - convert to warning with "Important:" prefix
  const patternImportant = /```\{important\}\n/g;
  if (patternImportant.test(content)) {
    modifiedContent = modifiedContent.replace(patternImportant, '```{warning}\n:class: important\n\n**Important:** ');
    changesMade.push('important → warning with Important prefix');
  }

  if (modifiedContent !== originalContent) {
    fs.writeFileSync(filepath, modifiedContent, 'utf8');
    return [true, changesMade];
  }
  return [false, []];
}

function main() {
  const contentDir = path.join(process.cwd(), 'content');

  if (!fs.existsSync(contentDir)) {
    console.log(`Error: Content directory not found: ${contentDir}`);
    process.exit(1);
  }

  // Find all markdown files recursively
  const markdownFiles = [];

  function walkDir(dir) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walkDir(fullPath);
      } else if (entry.name.endsWith('.md')) {
        markdownFiles.push(fullPath);
      }
    }
  }

  walkDir(contentDir);

  console.log(`Found ${markdownFiles.length} markdown files`);
  console.log('='.repeat(60));

  let totalChanged = 0;
  for (const filepath of markdownFiles) {
    const [changed, changes] = fixAdmonitionsInFile(filepath);
    if (changed) {
      totalChanged++;
      const relPath = path.relative(contentDir, filepath);
      console.log(`✓ ${relPath}`);
      for (const change of changes) {
        console.log(`  - ${change}`);
      }
    }
  }

  console.log('='.repeat(60));
  console.log(`Modified ${totalChanged} files`);

  if (totalChanged === 0) {
    console.log('No changes needed!');
  } else {
    console.log('\nAdmonitions have been converted to supported types.');
    console.log('Run \'npx myst build --typst\' to verify the fixes.');
  }
}

if (require.main === module) {
  main();
}

module.exports = { fixAdmonitionsInFile };
